### PR TITLE
feat: introduce let and const earlier

### LINKED ISSUE #43484: Basic JavaScript: course lesson could accept "let" or "const" in addition to "var"
## Affected page

https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/basic-javascript/finding-a-remainder-in-javascript

## Your code

```
let remainder;
remainder = 11 % 3;
```

## Expected behavior

Expected behavior would be for the test to pass, since the `remainder` variable has been initialized correctly. This follows the example of other early Basic JavaScript lessons that accept the `let` keyword when initializing variables.

Actual behavior: 2/3 tests pass, however test 1 fails with the message "The variable `remainder` should be initialized"

## System

- Device: MacBook Air (2017)
- OS: macOS Big Sur
- Browser: Chrome
- Version: 93.0.4577.82

## Additional context

This is a suggestion to update this particular lesson and any other early Basic JavaScript lessons that only accept `var` to initialize a variable. 

This is my first open source issue suggestion ever, so I am open to correction if I submitted this erroneously. Thank you for reading.

### PR DESCRIPTION
**Pull request by scissorsneedfoodtoo**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

This PR moves a couple of challenges from ES6 to an early portion of Basic JavaScript. The examples, text, and tests, "Compare Scopes of the var and let Keywords" and "Declare a Read-Only Variable with the const Keyword", had to be reworked to fit early in the curriculum.

A couple of other challenges were moved within the Basic JS section, but those are relatively minor changes, like moving "Declare a String Variable" up before learners have to actually work with strings.

Also, every challenge after `let` and `const` are introduced was reworked to use those keywords.

Note: These changes **should NOT be merged** until we decide how to redirect the old ES6 `let` and `const` challenges to their new locations in Basic JS:
- /learn/javascript-algorithms-and-data-structures/es6/explore-differences-between-the-var-and-let-keywords --> /learn/javascript-algorithms-and-data-structures/basic-javascript/explore-differences-between-the-var-and-let-keywords
- /learn/javascript-algorithms-and-data-structures/es6/declare-a-read-only-variable-with-the-const-keyword --> /learn/javascript-algorithms-and-data-structures/basic-javascript/declare-a-read-only-variable-with-the-const-keyword

Closes #43484

### CODE REVIEW

**Review by jatin837 — DISMISSED** [review]

**Comment by nhcarrigan** [comment]
@scissorsneedfoodtoo is this one still blocked?

**Comment by scissorsneedfoodtoo** [comment]
Hey @nhcarrigan, this is ready for review, just not set to be merged until we decide how to handle the redirects for the two ES6 challenges that were moved to Basic JS.

**Review by ShaunSHamilton — COMMENTED** [review]
We might want to change the descriptions of the intros as well: https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/i18n/locales/english/intro.json#L102

**Comment by scissorsneedfoodtoo** [comment]
Awesome, thanks for catching that @ShaunSHamilton. I'll work on that now.

**Comment by ojeytonwilliams** [comment]
@scissorsneedfoodtoo until we figure out a better way, the redirects are being handled by `serve`.  For example: https://github.com/freeCodeCamp/client-config/pull/4

@raisedadead are we being strict about keeping all old links alive?  We can definitely do it for these, but I wanted to check with you since the config will inevitably grow.

**Comment by raisedadead** [comment]
Yup, we want to preserve any links out there for SEO purposes. A 404 is pretty bad. I know this is pushing the issue with maintenance down the lane, but it's one of those things we have to live with. Please proceed with setting up redirects with serve for now. Once we have NGINX sitting in front of the web client (someday in future) it should be a non-issue (sort of).

**Comment by scissorsneedfoodtoo** [comment]
Thanks for the clarification and that link to the client redirects, @ojeytonwilliams.

I'll go ahead and add / test redirects for the two ES6 challenges we're moving.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/finding-a-remainder-in-javascript.md
Comment (ShaunSHamilton) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/finding-a-remainder-in-javascript.md:34]:
```suggestion
assert(/(const|let|var)\s+?remainder/.test(code));
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/global-scope-and-functions.md
Comment (ShaunSHamilton) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/global-scope-and-functions.md:39]:
```suggestion
assert(/(let|const)\s+myGlobal/.test(code));
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/iterate-through-an-array-with-a-for-loop.md
Comment (ShaunSHamilton) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/iterate-through-an-array-with-a-for-loop.md:66]: Just because I am here
```suggestion
const myArr = [2, 3, 4, 5, 6];
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/iterate-through-an-array-with-a-for-loop.md
Comment (ShaunSHamilton) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/iterate-through-an-array-with-a-for-loop.md:75]:
```suggestion
const myArr = [2, 3, 4, 5, 6];
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/implement-map-on-a-prototype.md
Comment (ShaunSHamilton) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/functional-programming/implement-map-on-a-prototype.md:63]:
```suggestion
  for (const elem of this) {
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
Some of this is necessary, others just annoyed me.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/concatenating-strings-with-plus-operator.md
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/concatenating-strings-with-plus-operator.md:53]: Thanks for catching this @ShaunSHamilton.

I played around with this for a bit, and it seems like the easiest thing to do is to change the seed to `let myStr;`, similar to the original version of the challenge.

Briefly considered `const myStr;`, but that would lead to an error when the page loads, and I'm not sure how the automated tests would handle that.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/declare-a-read-only-variable-with-the-const-keyword.md
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/declare-a-read-only-variable-with-the-const-keyword.md:71]: Awesome! The new test looks great, and your suggestions are much easier to reason about. Will commit this now. 

The only thing I'll go back and add are the comments in the seed, just on the off chance that someone tries to do this and runs into trouble:

```js
const FCC = "freeCodeCamp";
let fact = "is cool!";

let fact = "is awesome!";
console.log(FCC, fact);
```

**Review by ShaunSHamilton — DISMISSED** [review]
This LGTM 👍 . Nice work, Kris.

**Comment by ojeytonwilliams** [comment]
Nice job, @scissorsneedfoodtoo!

Before we get this in, I think we should go through the same process as https://github.com/freeCodeCamp/freeCodeCamp/pull/42596#issuecomment-897485766

- [ ] create redirect tests in this PR (which should fail)
- [ ] create a client config PR to fix the tests

when we're ready to deploy we

- [ ] merge the client config PR
- [ ] re-run the tests
- [ ] merge this
- [ ] deploy
- [ ] update the client config in the client VMs

Unfortunately, that's all blocked until we can do the last step, which Mrugesh is currently looking into.

**Comment by scissorsneedfoodtoo** [comment]
Thanks again for the thorough review @ShaunSHamilton, and for the detailed breakdown @ojeytonwilliams.

And no worries, there's no real rush on this. I'll fix the conflicts in https://github.com/freeCodeCamp/client-config/pull/5 in the meantime.

**Comment by ojeytonwilliams** [comment]
Hey @scissorsneedfoodtoo @raisedadead are we okay to deploy this now?

**Comment by raisedadead** [comment]
Logistically, we do have a window open. I will defer to Kris, for confirming though.

**Comment by scissorsneedfoodtoo** [comment]
Sorry, just seeing this now.

Sure, I think we're good to go on this if we still have an open window. I don't think there's anything else that needs to change here.

**Comment by scissorsneedfoodtoo** [comment]
Sorry again for the delay and force push. Saw that the directory structure for the Cypress tests changed a bit, so I thought it would be best to rebase.

The only change is that I added `cypress/integration/legacy/redirects/es6-to-basic-javascript.js` with a couple of tests for the two ES6 challenges that are being moved to the Basic JavaScript section:

* /learn/javascript-algorithms-and-data-structures/es6/explore-differences-between-the-var-and-let-keywords --> /learn/javascript-algorithms-and-data-structures/basic-javascript/explore-differences-between-the-var-and-let-keywords
* /learn/javascript-algorithms-and-data-structures/es6/declare-a-read-only-variable-with-the-const-keyword --> /learn/javascript-algorithms-and-data-structures/basic-javascript/declare-a-read-only-variable-with-the-const-keyword

If those tests look good, we should be able to carry out the steps @ojeytonwilliams detailed in https://github.com/freeCodeCamp/freeCodeCamp/pull/43133#issuecomment-904411754.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/concatenating-strings-with-plus-operator.md
Comment (ShaunSHamilton) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/concatenating-strings-with-plus-operator.md:76]: I worry we will get the usual complaints about the instructions being in the test text, and not in the instruction body:
> `myStr` should be created using the `const` keyword.

Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/concatenating-strings-with-plus-operator.md:76]: Awesome, thanks for catching that, @ShaunSHamilton. I'll go back in and add that instruction to the instruction text.

Comment (ieahleen) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/concatenating-strings-with-plus-operator.md:76]: I vote for number 2

Comment (ShaunSHamilton) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/concatenating-strings-with-plus-operator.md:76]: @scissorsneedfoodtoo Both of those are valid solutions. I am leaning towards option **1**, because it is less wordy. Although, I might even leave the `const` test in there, because Campers have a habit of changing more than they are asked.

Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/concatenating-strings-with-plus-operator.md:76]: Awesome, thanks for the input @ieahleen and @ShaunSHamilton.

I'm also leaning toward option 1 because it's a bit closer to what we have now. The only issue with leaving the `const` test is that it will pass with just the seed code. This was brought up in an earlier review.

I'm fine with it, as the current version of the challenge works this way with the `var` test. But I could go either way with options 1 or 2.

Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/concatenating-strings-with-plus-operator.md]: I went ahead with option 1 because it's closer to what we currently have.

If anyone else chimes in and would rather go with option 2, I can go back in and change the seed and wording.

Otherwise, I think this is ready for another quick review of the "Concatenating Strings with Plus Operator" challenge and for the steps Oliver mentioned in this comment above: https://github.com/freeCodeCamp/freeCodeCamp/pull/43133#issuecomment-904411754

**Review by ShaunSHamilton — DISMISSED** [review]
This LGTM 👍 Pending redirects

**Comment by scissorsneedfoodtoo** [comment]
@ieahleen, thanks for the ping about this.

The PR for the redirects is up here: https://github.com/freeCodeCamp/client-config/pull/5

I believe we're just waiting for another opportunity to follow the steps listed in [this comment](https://github.com/freeCodeCamp/freeCodeCamp/pull/43133#issuecomment-904411754). My understanding is that it'll take a bit of coordination because of that last step, "update the client config in the client VMs".

Is that right, @ojeytonwilliams? Has that bit been figured out?

It also might be worth merging this and https://github.com/freeCodeCamp/freeCodeCamp/pull/43429 at the same time, since they involve the same process listed above.

That PR seems to be missing a PR to handle the client redirects, though. I'll go ahead and create a redirect for those redirects now.

**Review by majestic-owl448 — APPROVED** [review]
LGTM :+1: pending tests

**Review by ShaunSHamilton — APPROVED** [review]

**Pull request merged**
